### PR TITLE
[MARKER]: (f)ALFF

### PR DESCRIPTION
**Pull request by fraimondo**
### Which marker do you want to include?

ALFF and fALFF

They can be calculated simultaneously as fALFF is just a scaling of ALFF in the full frequency range.

Inputs:
List of 3D MNI coordinates OR a mask file

Output:
Two arrays with one value per voxel, one for ALFF and one for fALFF

### Is there any publication or available code?

https://pubmed.ncbi.nlm.nih.gov/18501969/

### Do you have a sample code that implements this outside of junifer?

_No response_

### Anything else to say?

Equations here can be used:
https://web.conn-toolbox.org/fmri-methods/connectivity-measures/other#h.p_uBWGEgvizNhI

### CODE REVIEW

File: junifer/markers/falff/falff_base.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
+"""Provide abstract class for computing fALFF."""
```
Comment (synchon) [inline junifer/markers/falff/falff_base.py:1]: Please add a newline after this for consistency.

File: junifer/markers/falff/falff_base.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+
+class AmplitudeLowFrequencyFluctuationBase(BaseMarker):
+    """Base class for (fractional) Amplitude Low Frequency Fluctuation.
+
+    Parameters
+    ----------
+    fractional : bool
+        Whether to compute fractional ALFF.
+    highpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_base.py:22]: `highpass : positive float`?

File: junifer/markers/falff/falff_base.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        Highpass cutoff frequency.
+    lowpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_base.py:24]: `lowpass : positive float`?

File: junifer/markers/falff/falff_base.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        Lowpass cutoff frequency.
+    tr : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_base.py:26]: `tr : positive float, optional`?

File: junifer/markers/falff/falff_base.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        The Repetition Time of the BOLD data. If None, will extract
+        the TR from NIFTI header (default None).
+    use_afni : bool, optional
+        Whether to use AFNI for computing. If None, will use AFNI only
+        if available (default None).
+    name : str, optional
+        The name of the marker. If None, it will use the class name
+        (default None).
```
Comment (synchon) [inline junifer/markers/falff/falff_base.py:34]: Please add a newline after this.

File: junifer/markers/falff/falff_base.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        """Get valid data types for input.
+
+        Returns
+        -------
+        list of str
+            The list of data types that can be used as input for this marker.
+
+        """
+        valid = ["BOLD"]
+        return valid
```
Comment (synchon) [inline junifer/markers/falff/falff_base.py:90]: Maybe return the list directly?

File: junifer/markers/falff/falff_base.py (hunk 1 above)
Comment (synchon) [inline junifer/markers/falff/falff_base.py:1]: I don't think the class is abstract.

File: junifer/markers/falff/falff_base.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+    def _postprocess(self, input: Dict) -> Dict:
+        """Postprocess the output of the estimator.
+
+        Parameters
+        ----------
+        input : dict
+            The output of the estimator. It must have the following
+        """
+        raise NotImplementedError("_postprocess must be implemented")
```
Comment (synchon) [inline junifer/markers/falff/falff_base.py:174]: Use `raise_error`?

File: junifer/markers/falff/falff_estimator.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
+"""Provide estimator class for regional homogeneity (ReHo)."""
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:1]: The docstring needs to be updated.

File: junifer/markers/falff/falff_estimator.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+        lowpass: float,
+        tr: Optional[float],
+    ) -> Tuple["Nifti1Image", "Nifti1Image"]:
+        """Compute ALFF map via afni's commands.
+
+        Parameters
+        ----------
+        data : 4D Niimg-like object
+            Images to process.
+        highpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:110]: `highpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+            Highpass cutoff frequency.
+        lowpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:112]: `lowpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+        lowpass: float,
+        tr: Optional[float],
+    ) -> Tuple["Nifti1Image", "Nifti1Image"]:
+        """Compute (f)ALFF map.
+
+        Parameters
+        ----------
+        data : 4D Niimg-like object
+            Images to process.
+        highpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:191]: `highpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+            Highpass cutoff frequency.
+        lowpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:193]: `lowpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+            Lowpass cutoff frequency.
+        tr : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:114]: `tr : positive float, optional`

File: junifer/markers/falff/falff_estimator.py [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:195]: `tr : positive float, optional`?

File: junifer/markers/falff/falff_estimator.py [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+    ) -> Tuple["Nifti1Image", "Nifti1Image"]:
+        """Compute the ALFF map with memorization.
+
+        Parameters
+        ----------
+        use_afni : bool
+            Whether to use AFNI for computing.
+        data : 4D Niimg-like object
+            Images to process.
+        highpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:260]: `highpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+            Highpass cutoff frequency.
+        lowpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:262]: `lowpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+            Lowpass cutoff frequency.
+        tr : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:264]: `tr : positive float, optional`?

File: junifer/markers/falff/falff_estimator.py [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+    ) -> Tuple["Nifti1Image", "Nifti1Image"]:
+        """Fit and transform for the estimator.
+
+        Parameters
+        ----------
+        use_afni : bool
+            Whether to use AFNI for computing.
+        input_data : dict
+            The BOLD data as dictionary.
+        highpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:303]: `highpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 19]
```diff
@@ -0,0 +1,338 @@
...
+            Highpass cutoff frequency.
+        lowpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:305]: `lowpass : positive float`?

File: junifer/markers/falff/falff_estimator.py [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+            Lowpass cutoff frequency.
+        tr : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_estimator.py:307]: `tr : positive float, optional`?

File: junifer/markers/falff/falff_parcels.py [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    """Class for computing fALFF/ALFF on parcels.
+
+    Parameters
+    ----------
+    parcellation : str or list of str
+        The name(s) of the parcellation(s). Check valid options by calling
+        :func:`junifer.data.parcellations.list_parcellations`.
+    fractional : bool
+        Whether to compute fractional ALFF.
+    highpass : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_parcels.py:29]: `highpass : positive float, optional`?

File: junifer/markers/falff/falff_parcels.py [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+        The highpass cutoff frequency for the bandpass filter (default 0.01).
+    lowpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_parcels.py:31]: `lowpass : positive float, optional`?

File: junifer/markers/falff/falff_parcels.py [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+        The lowpass cutoff frequency for the bandpass filter (default 0.1).
+    tr : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_parcels.py:33]: `tr : positive float, optional`?

File: junifer/markers/falff/falff_parcels.py [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+    method_params : dict, optional
+        Parameters to pass to the aggregation function. Check valid options in
+        :func:`junifer.stats.get_aggfunc_by_name`.
+    name : str, optional
+        The name of the marker. If None, will use the class name (default
+        None).
+
+    Notes
+    -----
+    The `tr` parameter is crucial for the correctness of fALFF/ALFF
```
Comment (synchon) [inline junifer/markers/falff/falff_parcels.py:55]:
```
``tr``
```

File: junifer/markers/falff/falff_parcels.py [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+    computation. If a dataset is correctly preprocessed, the TR should be
+    extracted from the NIFTI without any issue. However, it has been
+    reported that some preprocessed data might not have the correct TR in
+    the NIFTI header.
+
+    ALFF/fALFF are computed using a bandpass butterworth filter. See
+    :func:`scipy.signal.butter` and :func:`scipy.signal.filtfilt` for more
+    details.
+
```
Comment (synchon) [inline junifer/markers/falff/falff_parcels.py:64]: Extra newline here, can be removed.

File: junifer/markers/falff/falff_spheres.py [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+    coords : str
+        The name of the coordinates list to use. See
+        :func:`junifer.data.coordinates.list_coordinates` for options.
+    radius : float, optional
+        The radius of the sphere in mm. If None, the signal will be extracted
+        from a single voxel. See :class:`nilearn.maskers.NiftiSpheresMasker`
+        for more information (default None).
+    fractional : bool
+        Whether to compute fractional ALFF.
+    highpass : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_spheres.py:33]: `highpass : positive float, optional`?

File: junifer/markers/falff/falff_spheres.py [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+        The highpass cutoff frequency for the bandpass filter (default 0.01).
+    lowpass : float
```
Comment (synchon) [inline junifer/markers/falff/falff_spheres.py:35]: `lowpass : positive float, optional`?

File: junifer/markers/falff/falff_spheres.py [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+        The lowpass cutoff frequency for the bandpass filter (default 0.1).
+    tr : float, optional
```
Comment (synchon) [inline junifer/markers/falff/falff_spheres.py:37]: `tr : positive float, optional`?

File: junifer/markers/falff/falff_spheres.py [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+    The `tr` parameter is crucial for the correctness of fALFF/ALFF
+    computation. If a dataset is correctly preprocessed, the TR should be
+    extracted from the NIFTI without any issue. However, it has been
+    reported that some preprocessed data might not have the correct TR in
+    the NIFTI header.
+
+    ALFF/fALFF are computed using a bandpass butterworth filter. See
+    :func:`scipy.signal.butter` and :func:`scipy.signal.filtfilt` for more
+    details.
+
```
Comment (synchon) [inline junifer/markers/falff/falff_spheres.py:68]: Extra newline here, can be removed.

File: junifer/markers/falff/falff_spheres.py [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+            confounds if available (default None).
+
+        Returns
+        -------
+        dict
+            The computed ALFF as dictionary. The dictionary has the following
+            keys:
+
+            * ``data`` : the actual computed values as a numpy.ndarray
+            * ``columns`` : the column labels for the computed values as a list
```
Comment (synchon) [inline junifer/markers/falff/falff_spheres.py:121]: Needs a newline after the listing.

File: junifer/markers/falff/tests/test_falff_estimator.py [hunk 31]
```diff
@@ -0,0 +1,241 @@
...
+    assert fifth > (first_time / 10)
+    n_files = len([x for x in estimator.temp_dir_path.glob("*")])
+    assert n_files == 3  # input + alff + falff
+
+
+@pytest.mark.skipif(
+    _check_afni() is False, reason="requires afni to be in PATH"
+)
+def test_AmplitudeLowFrequencyFluctuationEstimator_afni_vs_python() -> None:
+    """Test that the cache works properly when using afni."""
```
Comment (synchon) [inline junifer/markers/falff/tests/test_falff_estimator.py:209]: Docstring needs to be updated.

File: junifer/markers/falff/tests/test_falff_parcels.py [hunk 32]
```diff
@@ -0,0 +1,154 @@
...
+from scipy.stats import pearsonr
+
+from junifer.datareader import DefaultDataReader
+from junifer.markers.falff import AmplitudeLowFrequencyFluctuationParcels
+from junifer.testing.datagrabbers import PartlyCloudyTestingDataGrabber
+from junifer.pipeline.utils import _check_afni
+from junifer.storage import SQLiteFeatureStorage
+from junifer.utils import logger
+
+
```
Comment (synchon) [inline junifer/markers/falff/tests/test_falff_parcels.py:20]: As the parcellations used are same, maybe create a module-level variable and reuse it.

File: junifer/markers/falff/tests/test_falff_spheres.py [hunk 33]
```diff
@@ -0,0 +1,162 @@
+"""Provide test for parcel-aggregated (f)ALFF."""
```
Comment (synchon) [inline junifer/markers/falff/tests/test_falff_spheres.py:1]: Docstring needs to be updated.

File: junifer/markers/falff/tests/test_falff_spheres.py [hunk 34]
```diff
@@ -0,0 +1,162 @@
...
+from scipy.stats import pearsonr
+
+from junifer.datareader import DefaultDataReader
+from junifer.markers.falff import AmplitudeLowFrequencyFluctuationSpheres
+from junifer.testing.datagrabbers import PartlyCloudyTestingDataGrabber
+from junifer.pipeline.utils import _check_afni
+from junifer.storage import SQLiteFeatureStorage
+from junifer.utils import logger
+
+
```
Comment (synchon) [inline junifer/markers/falff/tests/test_falff_spheres.py:21]: As the coordinates used are same, maybe create a module-level variable and reuse it.

**Review by synchon — CHANGES_REQUESTED** [review]

File: junifer/markers/falff/falff_base.py (hunk 5 above)
Comment (fraimondo) [reply junifer/markers/falff/falff_base.py:34]: done

**Review by synchon — APPROVED** [review]

**Review by synchon — APPROVED** [review]

**Review by synchon — APPROVED** [review]

**Pull request merged**